File: backend/app/tolerance.py
"""Python mirror of shared/schema/src/tolerance.ts.

Keep behaviour consistent with the TypeScript implementation: the backend
recomputes every delivery's EFD and rejects submissions whose client-computed
values disagree.

    EFD = (VFD - VREF) / VREF * 100   [%]

VFD = volume indicated by the dispenser, VREF = volume indicated by the
reference measure. A delivery passes when |EFD| <= MPE.

MPE CONFIRMED (owner, 2026-07-31): SANS TEST PROC02 rev 14, Table 1
(LM-IR 117-1), accuracy class 0.5, Line A = 0.5 percent, the cell
highlighted as applicable in Prowalco's controlled procedure.
"""
import math
from dataclasses import dataclass

# Maximum permissible error for a fuel-dispenser delivery, in percent.
MPE_PERCENT = 0.5

# The delivery test points on the Metrologist Note, in report order.
DELIVERY_POINTS = ("del1_max", "del2_max", "del3_max", "min_flow_20l", "min_flow", "preset")

# Match the TS EPSILON nudge in roundTo(): Python's round() uses banker's
# rounding, so implement half-up rounding with the same epsilon behaviour.
_EPSILON = 2.220446049250313e-16


def round_to(value: float, decimals: int) -> float:
    # JS Math.round(x) === floor(x + 0.5) for all x, including negatives —
    # use the same formula so both implementations agree at .5 boundaries.
    f = 10.0**decimals
    return math.floor((value + _EPSILON) * f + 0.5) / f


@dataclass(frozen=True)
class EfdComputation:
    efd_percent: float
    passed: bool


def compute_efd(vfd_ml: float, vref_ml: float) -> EfdComputation:
    if vref_ml <= 0:
        raise ValueError("Reference volume (VREF) must be > 0")
    efd_percent = round_to(((vfd_ml - vref_ml) / vref_ml) * 100, 2)
    return EfdComputation(
        efd_percent=efd_percent,
        passed=abs(efd_percent) <= MPE_PERCENT,
    )

File: shared/schema/src/tolerance.ts
/**
 * Accuracy math for NRCS liquid-fuel-dispenser verification.
 *
 * The certificate reports, per delivery, the **EFD** (relative error of the
 * dispenser against the reference measure):
 *
 *     EFD = (VFD − VREF) / VREF × 100   [%]
 *
 * where VFD = volume indicated by the dispenser (LFD) and VREF = volume
 * indicated by the reference proving measure. A delivery passes when
 * |EFD| ≤ MPE.
 *
 * IMPORTANT: the MPE below is provisional. Confirm the exact NRCS / LM-IR
 * 117-2 maximum permissible error and the EFD sign convention with Prowalco's
 * quality manager before production use (see CLAUDE.md open questions).
 *
 * This module is mirrored in Python at backend/app/tolerance.py — keep the two
 * implementations consistent in behaviour. The backend recomputes every
 * delivery and rejects submissions whose client-computed EFD/outcome disagree.
 */

/** Maximum permissible error for a fuel-dispenser delivery, in percent.
 * PROVISIONAL — confirm with NRCS/QM. */
export const MPE_PERCENT = 0.5;

/** The delivery test points on the Metrologist Note, in report order. */
export const DELIVERY_POINTS = [
  'del1_max',
  'del2_max',
  'del3_max',
  'min_flow',
  'preset',
] as const;
export type DeliveryPoint = (typeof DELIVERY_POINTS)[number];

export const DELIVERY_POINT_LABELS: Record<DeliveryPoint, string> = {
  del1_max: 'Delivery 1 at max. achievable flow rate',
  del2_max: 'Delivery 2 at max. achievable flow rate',
  del3_max: 'Delivery 3 at max. achievable flow rate',
  min_flow: 'Delivery at minimum flow rate',
  preset: 'Preset delivery',
};

export interface EfdComputation {
  /** (VFD − VREF) / VREF × 100, rounded to 2 decimal places (%). */
  efdPercent: number;
  /** true when |EFD| ≤ MPE. */
  pass: boolean;
}

export function roundTo(value: number, decimals: number): number {
  const f = Math.pow(10, decimals);
  // EPSILON nudge avoids 0.5-boundary floating point misrounds (e.g. 1.005)
  return Math.round((value + Number.EPSILON) * f) / f;
}

/**
 * Compute the EFD (%) and pass/fail for one delivery.
 * @param vfdMl  volume indicated by the dispenser (mL or L — units cancel)
 * @param vrefMl volume indicated by the reference measure (same unit as vfd)
 */
export function computeEfd(vfdMl: number, vrefMl: number): EfdComputation {
  if (vrefMl <= 0) {
    throw new Error('Reference volume (VREF) must be > 0');
  }
  const efdPercent = roundTo(((vfdMl - vrefMl) / vrefMl) * 100, 2);
  return {
    efdPercent,
    pass: Math.abs(efdPercent) <= MPE_PERCENT,
  };
}
